### ISSUE TITLE
Prettier comments

### ISSUE THREAD

**Issue by amannn** [jlongster/prettier#265]
This is probably not the most important issue, but I'll submit it anyway in case someone finds this interesting.

Besides statements in code, I think it's also nice when comments are formatted in a way that makes them easy to read. I once hacked together a small atom plugin, that worked for my usecase: [format-javascript-comment](https://github.com/amannn/atom-format-javascript-comment).

Features that could be interesting for prettier are the following:

- It wraps lines at a specific line length by taking [minimum raggedness](https://en.wikipedia.org/wiki/Line_wrap_and_word_wrap#Minimum_raggedness) into account.
 - It wraps lines before JSDoc tags.

Here's a small demo:

<img src="https://raw.githubusercontent.com/amannn/atom-format-javascript-comment/master/usage.gif" />

It would of course be really cool if this was baked into prettier 🙂.

**Comment by jlongster** [comment]
This is a neat idea, but we need to make comments fully robust first. Currently they are moved around incorrectly and I'm about to rehaul how comments work. Let's get that in first and then we can talk about it.

I'm a little hesitant to do this because users may put big diagrams or something in comments, and like string templates, it's kind of nice that we don't touch them. Also, users can combine your tool with ours to format both separately. We can keep this issue open for now though.

Tweaking the title to make it more clear.

**Comment by sophistifunk** [comment]
I personally would be very wary of reformatting within comments, as it tends to break a lot of things such as ascii diagrams, table-formatted text, and the like.

**Comment by liddellj** [comment]
What about introducing a `wrapComments` option, defaulted to `false`? Although I can see how taking this approach in general could lead to an explosion of options...

**Comment by erbridge** [comment]
Or just allow the same flag to skip a comment block from being formatted as other code (pending in #671)?

**Comment by rattrayalex** [comment]
There was a PR adding this functionality to JSX not too long ago. Comments have also largely had the overhaul mentioned, but still need some work. 

If anyone is looking for a fun challenge, this may be a prime candidate!

**Comment by amannn** [comment]
Great to see some people interested in this idea!

@rattrayalex What would you suggest how this could be integrated? With or without an option? Having it be skippable with `prettier-ignore` is probably pretty important.

Btw. I recently discovered something my atom plugin currently doesn't cover: Treating code blocks like `{thisIsAProperty: true, hereIsAnotherProperty: false}` in comments atomic, meaning they can't be split by white space. I usually mark these parts with back ticks, so they should easily be identifiable (the back ticks are removed from my snippet here in Github).

**Comment by rattrayalex** [comment]
@amannn it would be great if this could be built well enough that it doesn't need to be an option. At a minimum, that means respecting `prettier-ignore` as you mentioned. It also means developing good heuristic around things like ascii-text, code examples (as you mentioned), etc. 

Probably the best place to start would be gathering a bunch of examples of comment blocks from around github, and forming a test file with them. 

It's also fantastic that you have an existing atom plugin; building on work & iteration from that would obviously be quite helpful.

**Comment by azz** [comment]
Imagine if we could format JavaScript inside Markdown (#2444) inside JavaScript comments!

```js
/**
 * ```js
 * if (hasScienceGoneTooFar()) {
 *   throw new Issue("prettier/prettier", "wow");
 * }
 * ```
 */
function wow() {}

**Comment by samhunta** [comment]
An issue with making wrapped comments respect `// prettier-ignore` is if the programmer's intent is to ignore the statement following the API comments it's ignoring. Instead there could be prettier-ignore-comment?

**Comment by dahjelle** [comment]
I just ran across a VS Code extension called [Rewrap](https://github.com/stkb/Rewrap) that does comment rewrapping. I've only been using it for a little while, but so far so good.

I agree that functionality like this would be fantastic within Prettier.

**Comment by suchipi** [comment]
This might be worth revisiting now; I think between the `fill` doc primitive and the markdown work, we have what we need. However, I expect there to be *tons* of issues where people are using comments in a special way. Maybe instead of formatting comments automatically, prettier should provide a way for the user to manually reformat a specific comment, instead of trying to reformat every comment? Like "reformat the comment at this row and column". Then editor integrations could call that. That way people could get nice comment formatting without us unintentionally breaking tons of comment-based stuff.

**Comment by rattrayalex** [comment]
Maybe a global `wrap-comments` option, and a `prettier-ignore-comment` directive that can be included in any comment?

Codebases that generally do weird things in comments can turn off the global flag, and codebases that generally don't can turn off the few places where wrapping is unhelpful...

thoughts?

**Comment by suchipi** [comment]
That would probably work, but I have concerns about the UX if the user has to use `prettier-ignore-comment` a lot. prettier-ignore is an escape hatch, so I'd hate it if the average user found themselves using it a ton. It creates friction where the user can get annoyed at the formatter instead of appreciating it. I think maybe we should implement a comment formatter and then run it on some large codebases and see how often we feel we need to use it. If we're using it a ton, we might need to iterate a bit.

**Comment by sorahn** [comment]
What if prettier just tried to format block comments `/** ... */` as markdown?

Or maybe you could even denote that it was a markdown comment. `/**md`

SublimeText (and others i'm sure) will switch languages when you do things like this too.

Example: 

```
$foo = <<<HTML
  <div />
HTML;
```

**Comment by amannn** [comment]
@suchipi I think you're right. After a first prototype we should be able to figure out how obtrusive this is and think of configuration options then.

I've been using my atom plugin for every comment that exceeds a 80 character line length for over 1,5 years now and I never ran into issues (even though it's very very basic). But for sure, code bases are different and I usually don't draw diagrams in comments 🙂. Maybe turning it on globally can work well for some people though.

In regards to escaping text: If we use the markdown formatter, maybe we could interpret blocks surrounded with triple back ticks as unknown and just ignore them. If there's explicitly a language specified, we can use prettier for the block if we support the given language. Not sure how the markdown formatter currently treats unknown code blocks?

I've read that blog article on the `fill` primitive. Am I right in the assumption that this doesn't consider [minimum raggedness](https://en.wikipedia.org/wiki/Line_wrap_and_word_wrap#Minimum_raggedness) but rather splits the line as soon as the next word wouldn't fit there? Personally I'd find it incredible if we'd go for minimum raggedness, but having `fill` (at least for the start) would be amazing for sure. Maybe the places where `fill` is currently used could also benefit from minimum raggedness though.

**Comment by suchipi** [comment]
Yeah; `fill` fills the line without respect to raggedness. I actually prefer that to the minimum raggedness algorithm personally (for code comments)- but it's a matter of preference.

**Comment by suchipi** [comment]
I used the [hard-wrap](https://atom.io/packages/hard-wrap) atom package for a long time up until recently when it broke with the latest version of atom. It behaves very similarly to fill.

**Comment by amannn** [comment]
Alright. I just had another observation.

I'm currently building a React component that looks like this:

```js
export default class GroupedStateProvider extends Component {
  static propTypes = {
    onSave: PropTypes.func.isRequired,
    render: PropTypes.func.isRequired,
    groups: PropTypes.object.isRequired, // `{[id]: {title: String, keys: [String]}}`
    values: PropTypes.object
  };
  
  ...

}
```

The comment that describes the shape of the `groups` property (I should really use a type system), is just a bit over 80 characters. Trailing comments could be a bit trickier, as they usually refer to the code on the left-hand side, so in this case the comment should be put on top in a new line and not wrapped.

Maybe it makes sense to limit the first implementation to standalone comments in order to reduce the scope.

**Comment by suchipi** [comment]
I have a lot of those kinds of comments for PropTypes.func, too. I agree that trailing comments are a bit different.

**Comment by Z3TA** [comment]
Idea: 
Comments would first be parsed as code (uncommented code is common)
But if there's too many syntax errors it should fall back to string/text

Could check for tabs and whitespace alignments and not break those, eg ascii graphs.

**Comment by sorahn** [comment]
@amannn `PropTypes.shape`?

**Comment by georgecrawford** [comment]
I'd imagine that it would be safe to assume that a long comment _without any line breaks_ could be formatted to add line breaks (at 80 chars but taking into account a ragged-right) without breaking anything. If there's a line break, Prettier could ignore it so that any ASCII art, etc. is preserved.

**Comment by suchipi** [comment]
@georgecrawford that could break `$FlowFixMe` and `eslint-disable-next-line` comments

**Comment by remcohaszing** [comment]
I would like to note that the ESLint [max-len](https://eslint.org/docs/rules/max-len#options) rule supports checking comments (not fixing).

Since this is an option of a rule that conflicts with `eslint-plugin-prettier` for other reasons, this means users can’t check this using ESLint either.

**Comment by j-f1** [comment]
The ESLint team created a [JSDoc parser](https://github.com/eslint/doctrine) which could be used to parse & reprint JSDoc comments (we currently have a way to identify JSDoc comments). I’m hoping to do this eventually, but it would be great if somebody were to step in and do this sooner!

**Comment by amannn** [comment]
@j-f1 Seems like [`doctrine`](https://github.com/eslint/doctrine), the JSDoc parser you mentioned above, [is no longer maintained](https://eslint.org/blog/2018/11/jsdoc-end-of-life).

However @gum3n recently [implemented a Prettier plugin to format JSDoc](https://github.com/prettier/prettier/issues/5659#issuecomment-589988440) which is based on a private fork of doctrine. Not sure what his perspective on this is?

I want to note that JSDoc is only part of this issue, the other aspect is simple comments (`//`) which can potentially span multiple lines.

**Comment by irekg** [comment]
My fork of doctrine have additional support for `yields` tag, nothing more. Mentioned plugin for JSDoc already wrap lines.

Just be aware that it doesn't respect prettier `printWidth` setting. It have it's own setting for that and columns are counted from first JSDoc character. So in practice if you have `printWidth` in prettier and `jsdocPrintWidth` in [prettier-plugin-jsdoc](https://gitlab.com/gumen/prettier-plugin-jsdoc) set to 80 the you indent JSDoc comment with 4 spaces then right boundary for JSDoc will be 84. It's because plugin is not written "in the right way" so it doesn't know about indentation.

To get familiar with features and limitations you can look at [test cases](https://gitlab.com/gumen/prettier-plugin-jsdoc/-/blob/master/tests/main.js).

**Comment by colecloudtostreet** [comment]
Looks like this has been open for 3 years. Is there a common workaround? or is everyone manually wrapping their comments like me?

**Comment by gum3n** [comment]
@colecloudtostreet Last comments talk about plugin https://gitlab.com/gumen/prettier-plugin-jsdoc that formats and wraps jsDoc comments. There is also version https://github.com/hosseinmd/prettier-plugin-jsdoc with more features and because it's based on first one it should have wrapping feature as well (its available on npm). Both plugins should have setting that allows to define comments length. Read previous comments for more details and try it :wink:

**Comment by felixfbecker** [comment]
It would be awesome if comments could be treated as markdown. That way they can reuse the existing Prettier formatter for markdown, can wrap prose inside it according to `proseWrap`, while ASCII  diagrams would stay intact as long as they are indented 4 spaces or wrapped in code fences, as well as bullet lists, tables, etc. There is a lot of precedent for this as VS Code treats docblocks as markdown and renders them as markdown in hover tooltips. All this would have to make sure is that any line starting with `@` would be treated as a special-case and not be prose-wrapped.

**Comment by hosseinmd** [comment]
https://github.com/hosseinmd/prettier-plugin-jsdoc is a plugin for format Jsdoc.

**Comment by muuvmuuv** [comment]
Yes, good idea to handle `/** */` as Markdown and IMO it would be easy if we would ignore  `/**/` or `//`. That way ASCII arts would be safe since that should use the `/**/` syntax anyway.

Additionally, separating blocks for easier reading should be considered as well: https://github.com/hosseinmd/prettier-plugin-jsdoc/issues/33

**Comment by cupidchan** [comment]
Is there any update on this wrap comments request?

**Comment by Sanaruca** [comment]
In the meantime I am using Rewrap, a VSCode extension.

**Comment by ph55** [comment]
> amannn opened this issue on Jan 17, 2017


![vVt9ZflEQplKefIfk4Ers6pbFFzvBPrPed_3k0f507U](https://github.com/user-attachments/assets/5a585ea0-d0a3-4473-983d-ab6692974b95)

**Comment by muuvmuuv** [comment]
> In the meantime I am using Rewrap, a VSCode extension.

Could that not be ported to an prettier plugin? Guess that would be a good thing for the time being.

**Comment by karlhorky** [comment]
It's F# and C# code - if anyone would be interested in starting to get involved in the Rewrap code, there's an initiative by the community to fork it here (original author is no longer around):

- https://github.com/stkb/Rewrap/issues/392

**Comment by muuvmuuv** [comment]
Porting to TS; Sounds like a job for ChatGPT 😅

**Comment by Silverbullet069** [comment]
Are there any progress regarding this feature?
